src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            checkIndexValidity(index);
+        }
+        if (index == 0 && size > 0) {
+            addToTheBeginning(value);
+        }
+        if (index == size) {
+            addToTheEnd(value);
+        } else if (index > 0) {
+            addToTheMiddle(value, index);
+        }
+    }
+
+    private void addToTheMiddle(T value, int index) {
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                Node<T> newNode = new Node<>(x.prev, value, x);
+                x.prev.next = newNode;
+                x.prev = newNode;
+                break;
+            }
+            i++;
+        }
+        size++;
+    }
+
+    private void addToTheEnd(T value) {
+        add(value);
+    }
+
+    private void addToTheBeginning(T value) {
+        Node<T> newNode = new Node<>(null, value, first);
+        first.prev = newNode;
+        first = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item: list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                return x.item;
+            }
+            i++;
+        }
         return null;
     }
 
+    private void checkIndexValidity(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
     @Override
     public T set(T value, int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                T oldValue = x.item;
+                x.item = value;
+                return oldValue;
+            }
+            i++;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                T removedItem = x.item;
+                remove(x.item);
+                return removedItem;
+            }
+            i++;
+        }
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (x.item == null) {
+                    unlink(x);
+                    return true;
+                }
+            }
+        } else {
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (object.equals(x.item)) {
+                    unlink(x);
+                    return true;
+                }
```

Comment:
You are iterating through the entire list to remove a specific object. This is inefficient, especially for large lists. You can optimize this by using a two-pointer approach: one pointer starts from the head (first node), the other from the tail (last node), and they meet in the middle.